File: android/app/src/main/java/com/onlycoffee/app/data/model/Coupon.kt
package com.onlycoffee.app.data.model

import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import kotlinx.parcelize.Parcelize
import java.text.SimpleDateFormat
import java.util.*

@Parcelize
data class Coupon(
    val id: String,
    @SerializedName("user_id")
    val userId: String,
    @SerializedName("promo_code_id")
    val promoCodeId: String? = null,
    val type: CouponType,
    val label: String,
    val description: String? = null,
    @SerializedName("value_cents")
    val valueCents: Int? = null,
    @SerializedName("percent_off")
    val percentOff: Int? = null,
    @SerializedName("price_override_cents")
    val priceOverrideCents: Int? = null,
    @SerializedName("eligible_items")
    val eligibleItems: EligibleItems? = null,
    val channels: String = "both",
    @SerializedName("expires_at")
    val expiresAt: Date,
    @SerializedName("redeemed_at")
    val redeemedAt: Date? = null,
    @SerializedName("redeemed_order_id")
    val redeemedOrderId: String? = null,
    val status: CouponStatus,
    val source: String,
    val metadata: Map<String, String>? = null,
    @SerializedName("created_at")
    val createdAt: Date
) : Parcelable {

    @Parcelize
    data class EligibleItems(
        val exclude: List<String>? = null,
        val include: List<String>? = null
    ) : Parcelable

    val isExpired: Boolean
        get() = Date().after(expiresAt)

    val isActive: Boolean
        get() = status == CouponStatus.ACTIVE && !isExpired

    val isExpiringSoon: Boolean
        get() {
            if (!isActive) return false
            val now = Date()
            val diffInMillis = expiresAt.time - now.time
            val hoursUntilExpiry = diffInMillis / (1000 * 60 * 60)
            return hoursUntilExpiry in 1..24
        }

    val displayValue: String
        get() = when (type) {
            CouponType.PERCENT_OFF -> "${percentOff}% OFF"
            CouponType.FIXED_PRICE -> {
                val price = (priceOverrideCents ?: 0) / 100.0
                String.format("$%.2f", price)
            }
            CouponType.FIXED_AMOUNT -> {
                val amount = (valueCents ?: 0) / 100.0
                String.format("$%.2f OFF", amount)
            }
            CouponType.FREE_ITEM -> "FREE"
        }

    val expiryText: String
        get() {
            val now = Date()
            if (now.after(expiresAt)) {
                return "Expired"
            }

            val diffInMillis = expiresAt.time - now.time
            val days = diffInMillis / (1000 * 60 * 60 * 24)
            val hours = (diffInMillis / (1000 * 60 * 60)) % 24

            return when {
                days > 1 -> "Expires in $days days"
                days == 1L -> "Expires in 1 day"
                hours > 1 -> "Expires in $hours hours"
                hours == 1L -> "Expires in 1 hour"
                else -> "Expires soon"
            }
        }

    val formattedExpiryDate: String
        get() {
            val format = SimpleDateFormat("MMM dd, yyyy 'at' hh:mm a", Locale.getDefault())
            return format.format(expiresAt)
        }

    val channelText: String
        get() = when (channels) {
            "app_only" -> "App Only"
            "store_only" -> "In-Store Only"
            else -> "App & In-Store"
        }
}

enum class CouponType {
    @SerializedName("percent_off")
    PERCENT_OFF,

    @SerializedName("fixed_price")
    FIXED_PRICE,

    @SerializedName("fixed_amount")
    FIXED_AMOUNT,

    @SerializedName("free_item")
    FREE_ITEM
}

enum class CouponStatus {
    @SerializedName("active")
    ACTIVE,

    @SerializedName("redeemed")
    REDEEMED,

    @SerializedName("expired")
    EXPIRED,

    @SerializedName("cancelled")
    CANCELLED
}

// API Response models
data class CouponsResponse(
    val success: Boolean,
    val data: CouponsData
)

data class CouponsData(
    val active: List<Coupon>,
    val expired: List<Coupon>
)

data class RedeemPromoCodeResponse(
    val success: Boolean,
    val message: String,
    val data: List<Coupon>?
)

data class RedeemPromoCodeRequest(
    val code: String,
    @SerializedName("idempotency_key")
    val idempotencyKey: String? = null
)

data class RedeemCouponRequest(
    @SerializedName("coupon_id")
    val couponId: String,
    @SerializedName("order_id")
    val orderId: String
)

data class CouponResponse(
    val success: Boolean,
    val data: Coupon
)

File: mobile/ios/OnlyCoffee/Models/Coupon.swift
import Foundation

// MARK: - Coupon Type Enum
enum CouponType: String, Codable {
    case percentOff = "percent_off"
    case fixedPrice = "fixed_price"
    case fixedAmount = "fixed_amount"
    case freeItem = "free_item"
}

// MARK: - Coupon Status Enum
enum CouponStatus: String, Codable {
    case active = "active"
    case redeemed = "redeemed"
    case expired = "expired"
    case cancelled = "cancelled"
}

// MARK: - Coupon Model
struct Coupon: Codable, Identifiable {
    let id: String
    let userId: String
    let promoCodeId: String?
    let type: CouponType
    let label: String
    let description: String?
    let valueCents: Int?
    let percentOff: Int?
    let priceOverrideCents: Int?
    let eligibleItems: EligibleItems?
    let channels: String
    let expiresAt: Date
    let redeemedAt: Date?
    let redeemedOrderId: String?
    let status: CouponStatus
    let source: String
    let metadata: [String: AnyCodable]?
    let createdAt: Date

    // MARK: - Eligible Items
    struct EligibleItems: Codable {
        let exclude: [String]?
        let include: [String]?
    }

    // MARK: - Computed Properties
    var isActive: Bool {
        status == .active && !isExpired
    }

    var isExpired: Bool {
        Date() >= expiresAt
    }

    var isExpiringSoon: Bool {
        guard isActive else { return false }
        let hoursUntilExpiry = Calendar.current.dateComponents([.hour], from: Date(), to: expiresAt).hour ?? 0
        return hoursUntilExpiry <= 24
    }

    var displayValue: String {
        switch type {
        case .percentOff:
            return "\(percentOff ?? 0)% OFF"
        case .fixedPrice:
            let price = Double(priceOverrideCents ?? 0) / 100.0
            return String(format: "$%.2f", price)
        case .fixedAmount:
            let amount = Double(valueCents ?? 0) / 100.0
            return String(format: "$%.2f OFF", amount)
        case .freeItem:
            return "FREE"
        }
    }

    var expiryText: String {
        let calendar = Calendar.current
        let now = Date()

        if isExpired {
            return "Expired"
        }

        let components = calendar.dateComponents([.day, .hour], from: now, to: expiresAt)

        if let days = components.day, days > 0 {
            return "Expires in \(days) day\(days == 1 ? "" : "s")"
        } else if let hours = components.hour, hours > 0 {
            return "Expires in \(hours) hour\(hours == 1 ? "" : "s")"
        } else {
            return "Expires soon"
        }
    }

    var channelText: String {
        switch channels {
        case "app_only":
            return "App Only"
        case "in_store":
            return "In-Store Only"
        default:
            return "App or In-Store"
        }
    }

    var formattedExpiryDate: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d, h:mm a"
        formatter.timeZone = TimeZone(identifier: "America/Chicago")
        return "Expires \(formatter.string(from: expiresAt))"
    }
}

// MARK: - AnyCodable Helper
struct AnyCodable: Codable {
    let value: Any

    init(_ value: Any) {
        self.value = value
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()

        if let bool = try? container.decode(Bool.self) {
            value = bool
        } else if let int = try? container.decode(Int.self) {
            value = int
        } else if let double = try? container.decode(Double.self) {
            value = double
        } else if let string = try? container.decode(String.self) {
            value = string
        } else if let array = try? container.decode([AnyCodable].self) {
            value = array.map { $0.value }
        } else if let dictionary = try? container.decode([String: AnyCodable].self) {
            value = dictionary.mapValues { $0.value }
        } else {
            value = NSNull()
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()

        switch value {
        case let bool as Bool:
            try container.encode(bool)
        case let int as Int:
            try container.encode(int)
        case let double as Double:
            try container.encode(double)
        case let string as String:
            try container.encode(string)
        case let array as [Any]:
            try container.encode(array.map { AnyCodable($0) })
        case let dictionary as [String: Any]:
            try container.encode(dictionary.mapValues { AnyCodable($0) })
        default:
            try container.encodeNil()
        }
    }
}

// MARK: - API Response Models
struct CouponsResponse: Codable {
    let success: Bool
    let data: CouponsData

    struct CouponsData: Codable {
        let active: [Coupon]
        let expired: [Coupon]
        let redeemed: [Coupon]
        let all: [Coupon]
    }
}

struct RedeemPromoCodeResponse: Codable {
    let success: Bool
    let message: String
    let data: [Coupon]?
}

struct CouponResponse: Codable {
    let success: Bool
    let data: Coupon
}
